server/src/main/java/io/pravega/schemaregistry/server/rest/resources/GroupResourceImpl.java

```diff
@@ -0,0 +1,690 @@
+import io.pravega.schemaregistry.exceptions.CodecTypeNotRegisteredException;
+import io.pravega.schemaregistry.exceptions.IncompatibleSchemaException;
+import io.pravega.schemaregistry.exceptions.PreconditionFailedException;
+import io.pravega.schemaregistry.exceptions.SerializationFormatMismatchException;
+import io.pravega.schemaregistry.server.rest.ServiceConfig;
+import io.pravega.schemaregistry.service.SchemaRegistryService;
+import io.pravega.schemaregistry.storage.ContinuationToken;
+import io.pravega.schemaregistry.storage.StoreExceptions;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.core.Response;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static io.pravega.auth.AuthHandler.Permissions.READ;
+import static io.pravega.auth.AuthHandler.Permissions.READ_UPDATE;
+import static javax.ws.rs.core.Response.Status;
+
+/**
+ * Schema Registry Resource implementation.
+ */
+@Slf4j
+public class GroupResourceImpl extends AbstractResource implements ApiV1.GroupsApiAsync {
+    private static final int DEFAULT_LIST_GROUPS_LIMIT = 100;
+    
+    public GroupResourceImpl(SchemaRegistryService registryService, ServiceConfig config, Executor executor) {
+        super(registryService, config, executor);
+    }
+
+    @Override
+    public void listGroups(String continuationToken, Integer limit, String namespace,
+                           AsyncResponse asyncResponse) {
+        log.info("List Groups called");
+        int toFetch = limit == null ? DEFAULT_LIST_GROUPS_LIMIT : limit;
+        ListGroupsResponse groupsList = new ListGroupsResponse();
+
+        List<String> authorizationHeader = getConfig().isAuthEnabled() ? getAuthorizationHeader() : Collections.emptyList();
+        CompletableFuture.runAsync(() -> {
+            if (getConfig().isAuthEnabled()) {
+                String credentials = parseCredentials(authorizationHeader);
+                try {
+                    getAuthManager().authenticate(credentials);
+                } catch (AuthException e) {
+                    log.warn("Unauthenticated", e);
+                    asyncResponse.resume(Response.status(Response.Status.FORBIDDEN.getStatusCode()).build());
+                }
+            }
+        }, getExecutorService()).thenCompose(v -> {
+            Predicate<Map.Entry<String, GroupProperties>> predicate = x -> {
+                try {
+                    String resource = Strings.isNullOrEmpty(namespace) ?
+                            getGroupResource(x.getKey()) :
+                            getGroupResource(x.getKey(), namespace);
+
+                    authenticateAuthorize(authorizationHeader, resource, READ);
```

Comment:
Authentication seems to be happening again unnecessarily for the second time. You might either want to only authorization here, or better still do both together in one place.